File: 005-quick-sort/quick-sort.js
function quickSort(a, p, r) {
  if (p < r) {
    q = randomPartition(a, p, r);
    quickSort(a, p, q - 1);
    quickSort(a, q + 1, r);
  }
}

function randomPartition(a, p, r) {
  var pivot = getRandom(p, r);
  var tmp = a[pivot];
  a[pivot] = a[r];
  a[r] = tmp;
  return partition(a, p, r);
}

function getRandom(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function partition(a, p, r) {
  var i = p - 1;
  for (var j = p; j < r; j++) {
    if (a[j] <= a[r]) {
      i = i + 1;
      var tmp = a[i];
      a[i] = a[j];
      a[j] = tmp;
    }
  }
  i = i + 1;
  var tmp = a[i];
  a[i] = a[r];
  a[r] = tmp;
  return i;
}

var a = [10, 9, 17, 23, 12, -2, 8, 45];
console.log("Before sorting", a);
quickSort(a, 0, a.length - 1);
console.log("After sorting", a);

File: 005-quick-sort/quick-sort.go
package main

import (
	"fmt"
	"math/rand"
)

func main() {
	a := []int{10, 9, 3, -1, 8, 4, 23, 98, 1, 34}
	fmt.Println("Before sorting", a)
	quickSort(a, 0, len(a)-1)
	fmt.Println("After sorting", a)
}

func quickSort(a []int, p int, r int) {
	if p < r {
		q := randomPartition(a, p, r)
		quickSort(a, p, q-1)
		quickSort(a, q+1, r)
	}
}

func randomPartition(a []int, p int, r int) int {
	pivot := getRandom(p, r)
	swap(&a[pivot], &a[r])
	return partition(a, p, r)
}

func getRandom(min int, max int) int {
	r := rand.New(rand.NewSource(55))
	return r.Intn(max-min+1) + min
}

func partition(a []int, p int, r int) int {
	i := p - 1
	for j := p; j < r; j++ {
		if a[j] <= a[r] {
			i = i + 1
			swap(&a[i], &a[j])
		}
	}
	i = i + 1
	swap(&a[i], &a[r])
	return i
}

func swap(x *int, y *int) {
	tmp := *x
	*x = *y
	*y = tmp
}
